Bug: Unexpected Verdicts for the Rule "no-control-regex"

### Environment

Node version: v16.13.0
npm version: v8.1.0
Local ESLint version: v8.9.0 (Currently used)
Global ESLint version: Not found
Operating System: darwin


### What parser are you using?

@typescript-eslint/parser

### What did you do?

<details>
<summary>Configuration</summary>

```
module.exports = {
	env: {
		browser: true,
		es2021: true,
		node: true
	},
	extends: [
		'eslint:recommended',
		'plugin:@typescript-eslint/recommended'
	],
	parser: '@typescript-eslint/parser',
	parserOptions: {
		ecmaVersion: 'latest',
		sourceType: 'module'
	},
	plugins: [
		'@typescript-eslint'
	],
	rules: {
	}
}
```
</details>

```js
const a = new RegExp('\t');
  // expected: error; actual: error
const b = new RegExp('\\t');
  // expected: no error; actual: no error
const c = new RegExp('\u0009');
  // expected: error; actual: error
const d = new RegExp('\\u0009');
  // expected: no error; actual: error
const e = new RegExp('\x09');
  // expected: error; actual: error
const f = new RegExp('\\x09');
  // expected: no error; actual: error
```


### What did you expect to happen?

I expected no errors for the lines defining `d` and `f`. The comments in the reproduction example show my expectations and the observed behavior.

### What actually happened?

I am seeing unexpected errors for the lines defining `d` and `f` (lines 7 and 11):

```
   1:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   5:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   7:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
   9:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
  11:22  error  Unexpected control character(s) in regular expression: \x09  no-control-regex
```

The errors for the lines 1, 5 and 9 are expected.

### Participation

- [ ] I am willing to submit a pull request for this issue.

### Additional comments

As a rationale of  the rule https://eslint.org/docs/rules/no-control-regex says: "These characters are rarely used in JavaScript strings so a regular expression containing these characters is most likely a mistake".

Note that the strings `'\\u0009'` and `'\\x09'` do not contain control characters: `'\\u0009' === String.fromCharCode(0x5C, 0x75, 0x30, 0x30, 0x30, 0x39)` and `'\\x09' === String.fromCharCode(0x5C, 0x78, 0x30, 0x39)`.

--- a/docs/src/rules/no-control-regex.md
+++ b/docs/src/rules/no-control-regex.md
@@ -19,8 +19,7 @@ The following elements of regular expression patterns are considered possible er
 
 * Hexadecimal character escapes from `\x00` to `\x1F`.
 * Unicode character escapes from `\u0000` to `\u001F`.
-* Raw characters from U+0000 to U+001F.
-* Escaped raw characters from U+0000 to U+001F.
+* Unescaped raw characters from U+0000 to U+001F.
 
 Control escapes such as `\t` and `\n` are allowed by this rule.
 
@@ -34,7 +33,7 @@ var pattern2 = /\x0C/;
 var pattern3 = /\x1F/;
 var pattern4 = /\u000C/;
 var pattern5 = new RegExp("\x0C"); // raw U+000C character in the pattern
-var pattern6 = new RegExp("\\x0C"); // escaped raw U+000C character in the pattern
+var pattern6 = new RegExp("\\x0C"); // \x0C pattern
 ```
 
 Examples of **correct** code for this rule:

--- a/docs/src/rules/no-control-regex.md
+++ b/docs/src/rules/no-control-regex.md
@@ -9,19 +9,32 @@ rule_type: problem
 
 Disallows control characters in regular expressions.
 
-Control characters are special, invisible characters in the ASCII range 0-31. These characters are rarely used in JavaScript strings so a regular expression containing these characters is most likely a mistake.
+Control characters are special, invisible characters in the ASCII range 0-31. These characters are rarely used in JavaScript strings so a regular expression containing elements that explicitly match these characters is most likely a mistake.
 
 ## Rule Details
 
-This rule disallows control characters in regular expressions.
+This rule disallows control characters and some escape sequences that match control characters in regular expressions.
+
+The following elements of regular expression patterns are considered possible errors in typing and are therefore disallowed by this rule:
+
+* Hexadecimal character escapes from `\x00` to `\x1F`.
+* Unicode character escapes from `\u0000` to `\u001F`.
+* Raw characters from U+0000 to U+001F.
+* Escaped raw characters from U+0000 to U+001F.
+
+Control escapes such as `\t` and `\n` are allowed by this rule.
 
 Examples of **incorrect** code for this rule:
 
 ```js
 /*eslint no-control-regex: "error"*/
 
-var pattern1 = /\x1f/;
-var pattern2 = new RegExp("\x1f");
+var pattern1 = /\x00/;
+var pattern2 = /\x0C/;
+var pattern3 = /\x1F/;
+var pattern4 = /\u000C/;
+var pattern5 = new RegExp("\x0C"); // raw U+000C character in the pattern
+var pattern6 = new RegExp("\\x0C"); // escaped raw U+000C character in the pattern
 ```
 
 Examples of **correct** code for this rule:
@@ -30,8 +43,27 @@ Examples of **correct** code for this rule:
 /*eslint no-control-regex: "error"*/
 
 var pattern1 = /\x20/;
-var pattern2 = new RegExp("\x20");
+var pattern2 = /\u0020/;
+var pattern3 = /\t/;
+var pattern4 = /\n/;
+var pattern5 = new RegExp("\x20");
+var pattern6 = new RegExp("\\t");
+var pattern7 = new RegExp("\\n");
 ```
+
+## Known Limitations
+
+When checking `RegExp` constructor calls, this rule examines evaluated regular expression patterns. Therefore, although this rule intends to allow syntax such as `\t`, it doesn't allow `new RegExp("\t")` since the evaluated pattern (string value of `"\t"`) contains a raw control character (the TAB character).
+
+```js
+/*eslint no-control-regex: "error"*/
+
+new RegExp("\t"); // disallowed since the pattern is: <TAB>
+
+new RegExp("\\t"); // allowed since the pattern is: \t
+```
+
+There is no difference in behavior between `new RegExp("\t")` and `new RegExp("\\t")`, and the intention to match the TAB character is clear in both cases. They are equally valid for the purpose of this rule, but it only allows `new RegExp("\\t")`.
 
 ## When Not To Use It
 

